realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -387,7 +388,12 @@ private void waitForChanges(int direction) throws InterruptedException {
 
                 throw new ObjectServerError(ErrorCode.UNKNOWN, errorMsg + " Has the SyncClient been started?");
             }
-            wrapper.waitForServerChanges();
+            try {
+                wrapper.waitForServerChanges();
+            } catch(InterruptedException e) {
+                waitingForServerChanges.set(null); // Ignore any results being sent if the wait was interrupted.
```

Comment:
So, canceling was neither supported in Object Store or Sync. So I introduced a callback Id instead that is used to determine if a notification should be used or not.